### ISSUE TITLE
Type variable replaced with `Unknown` in generic type

### ISSUE THREAD

**Issue by ethframe** [microsoft/pyright#2462]
**Describe the bug**
Type variable replaced with `Unknown` when it is used as an argument to a generic class type inside a method of that class. 

**To Reproduce**
```
$ pyright foo.py
...
...\bug.py
  ...\bug.py:12:21 - info: Type of "x" is "T@A"
  ...\bug.py:14:21 - info: Type of "t" is "tuple[T@A]"
  ...\bug.py:15:9 - error: Type of "a" is partially unknown
    Type of "a" is "A[Unknown]" (reportUnknownVariableType)
  ...\bug.py:16:21 - info: Type of "a" is "A[Unknown]"
  ...\bug.py:17:16 - error: Return type, "A[Unknown]", is partially unknown (reportUnknownVariableType)
  ...\bug.py:21:21 - info: Type of "x" is "T@A"
  ...\bug.py:23:21 - info: Type of "t" is "tuple[T@A]"
  ...\bug.py:25:21 - info: Type of "a" is "A[tuple[T@A]]"
2 errors, 0 warnings, 6 infos
```

**Expected behavior**
No errors, type of "a" is `A[T@A]` in the first method.

**Screenshots or Code**

```python
from typing import Generic, Tuple, TypeVar

T = TypeVar("T")


class A(Generic[T]):
    def __init__(self, x: T):
        self.x = x

    def a(self) -> 'A[T]':
        x = self.x
        reveal_type(x)  # T@A
        t = (x,)
        reveal_type(t)  # tuple[T@A]
        a = A(t[0])
        reveal_type(a)  # A[Unknown]
        return a  # Error

    def b(self) -> 'A[Tuple[T]]':
        x = self.x
        reveal_type(x)  # T@A
        t = (x,)
        reveal_type(t)  # tuple[T@A]
        a = A(t)
        reveal_type(a)  # A[tuple[T@A]]
        return a
```

**VS Code extension or command-line**
pyright 1.1.180

**Comment by ethframe** [comment]
I've found workarounds:

```python
from typing import Generic, TypeVar


T = TypeVar("T")


class A(Generic[T]):
    def __init__(self, x: T):
        self.x = x

    def a(self) -> "A[T]":
        a = A(self.x)
        return a  # Error

    def b(self) -> "A[T]":
        return A(self.x)  # Ok

    def c(self) -> "A[T]":
        a: A[T] = A(self.x)
        return a  # Ok
```

```
$ pyright bug.py
...
...\bug.py
  ...\bug.py:12:9 - error: Type of "a" is partially unknown
    Type of "a" is "A[Unknown]" (reportUnknownVariableType)
  ...\bug.py:13:16 - error: Return type, "A[Unknown]", is partially unknown (reportUnknownVariableType)
2 errors, 0 warnings, 0 infos
```

**Comment by erictraut** [comment]
This issue is related to another bug you filed earlier: https://github.com/microsoft/pyright/issues/2103. I've tried at least six ways to address that issue. Most of them didn't work (introduced regressions). The solution that I have in place is one I'm still not happy with — it's inelegant and still has the hole that you've identified here. I'm still thinking about a better way to handle it such that all of these cases work.

The problem occurs when you construct an instance of a class from within itself and parameterize it with a type that includes a type variable scoped to that class, such as `A[List[T]] = A()`. The `T` in the newly-constructed class is now distinct from the `T` in the class that constructed. Tracking the difference between these two `T` instances is challenging, at least given the current internal architectural.

What you're doing here is not illegal, but it's quite unusual. I'll continue to think about a better solution.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.193, which I just published. It will also be included in the next release of pylance.
